sdks/python/apache_beam/transforms/trigger_test.py

```diff
@@ -659,11 +675,132 @@ def fire_timers():
     self.assertEqual([], output, msg='Unexpected output: %s' % output)
 
 
+class TestStreamTranscriptTest(TranscriptTest):
+  """A suite of TestStream-based tests based on trigger transcript entries.
+  """
+
+  def _execute(
+      self, window_fn, trigger_fn, accumulation_mode, timestamp_combiner,
+      transcript, spec):
+
+    runner_name = TestPipeline().runner.__class__.__name__
+    if runner_name in spec.get('broken_on', ()):
+      self.skipTest('Known to be broken on %s' % runner_name)
+
+    test_stream = TestStream()
+    for action, params in transcript:
+      if action == 'expect':
+        test_stream.add_elements([('expect', params)])
+      else:
+        test_stream.add_elements([('expect', [])])
+        if action == 'input':
+          test_stream.add_elements([('input', e) for e in params])
+        elif action == 'watermark':
+          test_stream.advance_watermark_to(params)
+        elif action == 'clock':
+          test_stream.advance_processing_time(params)
+        elif action == 'state':
+          pass  # Requires inspection of implementation details.
+        else:
+          raise ValueError('Unexpected action: %s' % action)
+    test_stream.add_elements([('expect', [])])
+
+    class Check(beam.DoFn):
+      """A StatefulDoFn that verifies outputs are produced as expected.
+
+      This DoFn takes in two kinds of inputs, actual outputs and
+      expected outputs.  When an actual output is received, it is buffered
+      into state, and when an expected output is received, this buffered
+      state is retrieved and compared against the expected value(s) to ensure
+      they match.
+
+      The key is ignored, but all items must be on the same key to share state.
+      """
+      def process(
+          self, element, seen=beam.DoFn.StateParam(
+              beam.transforms.userstate.BagStateSpec(
+                  'seen',
+                  beam.coders.FastPrimitivesCoder()))):
+        _, (action, data) = element
+        if action == 'actual':
+          seen.add(data)
+
+        elif action == 'expect':
+          actual = list(seen.read())
+          seen.clear()
+
+          if len(actual) > len(data):
+            raise AssertionError(
+                'Unexpected output: expected %s but got %s' % (data, actual))
+          elif len(data) > len(actual):
+            raise AssertionError(
+                'Unmatched output: expected %s but got %s' % (data, actual))
```

Comment:
Nit: The elif branch does not seem to provide more information than the if branch. Shall we just merge them to one: if len(data) != len(actual)?

Reply:
The error strings are different, which are used to match the errors in the test.